### PR TITLE
fix: 8541: Resolve multiple transition.

### PR DESCRIPTION
**Pull request by dharmik**
### Fixed Issues

$ https://github.com/Expensify/App/issues/8541

### Tests

tests to something as below -

1 Login to New expensify
2 tap on profile
3 Tap on issue card
4 Tap on connect bank account

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose and it is
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn’t already exist
    - [x] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps

tests to something like this -

1 Login to New expensify
2 tap on profile
3 Tap on issue card
4 Tap on connect bank account

- [x] Verify that no errors appear in the JS console

### Video

iOS
https://user-images.githubusercontent.com/1510175/179397532-7dc20171-cee1-473a-ac10-900bb67262fa.mov

android
https://drive.google.com/file/d/1RbDU7nGZS_l4SUeyzt7ODePctWD7fnh5/view?usp=sharing

Desktop
https://drive.google.com/file/d/1rxk_eKyBRV3t0CvRaCH3WBeBCaHmDW3l/view?usp=sharing

Web
https://drive.google.com/file/d/1OskYEA2YRwNiApCv_d791yu_gRzkVlFW/view?usp=sharing

Mobile Web
https://drive.google.com/file/d/1JZDMy5PrY0LVKqmP9-HWOOrzgb_nn1tT/view?usp=sharing

### CODE REVIEW

File: src/pages/workspace/WorkSpaceUtils.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import lodashGet from 'lodash/get';
+import BankAccount from '../../libs/models/BankAccount';
+import Navigation from '../../libs/Navigation/Navigation';
+
+function getShouldShowPage(props, nextPage, defaultPage) {
+    const state = lodashGet(props.reimbursementAccount, 'achData.state');
+    const isShowPage = lodashGet(props.reimbursementAccount, 'achData.bankAccountID') && state !== +BankAccount.STATE.OPEN;
```
Comment (parasharrajat) [inline src/pages/workspace/WorkSpaceUtils.js:7]:
```suggestion
    const shouldShowPage = lodashGet(props.reimbursementAccount, 'achData.bankAccountID') && state !== +BankAccount.STATE.OPEN;
```

Why is this + in `+BankAccount.STATE.OPEN`?

File: src/pages/workspace/WorkSpaceUtils.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import lodashGet from 'lodash/get';
+import BankAccount from '../../libs/models/BankAccount';
+import Navigation from '../../libs/Navigation/Navigation';
+
+function getShouldShowPage(props, nextPage, defaultPage) {
```
Comment (parasharrajat) [inline src/pages/workspace/WorkSpaceUtils.js:5]: Name the should on what it does?

File: src/pages/workspace/WorkspaceInitialPage.js [hunk 3]
```diff
@@ -109,7 +110,7 @@ class WorkspaceInitialPage extends React.Component {
             {
                 translationKey: 'workspace.common.bankAccount',
                 icon: Expensicons.Bank,
-                action: () => Navigation.navigate(ROUTES.getWorkspaceBankAccountRoute(policy.id)),
+                action: () => WorkSpaceUtils.getShouldShowPage(this.props, ROUTES.getWorkspaceBankAccountRoute(policy.id), ROUTES.getBankAccountRoute()),
```
Comment (parasharrajat) [inline src/pages/workspace/WorkspaceInitialPage.js:113]: All of these statements have a lot of duplicate code. Shouldn't we move that inside the function? 

Also, do not pass `this.props`. Pass only necessary data to the function which is really needed.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Please fix the lint issue and add videos for all platforms. Improve QA steps and tests. They are also incomplete.

Fix the issue link. 
image

`$ https://github.com/Expensify/App/issues/8541`

File: src/pages/workspace/WorkspaceInitialPage.js (hunk 3 above)
Comment (dharmik) [reply src/pages/workspace/WorkspaceInitialPage.js:125]: Done

File: src/pages/workspace/WorkSpaceUtils.js (hunk 1 above)
Comment (dharmik) [reply src/pages/workspace/WorkSpaceUtils.js]: Removed.

File: src/pages/workspace/WorkSpaceUtils.js (hunk 2 above)
Comment (dharmik) [reply src/pages/workspace/WorkSpaceUtils.js]: Changed. I change with `getShouldShowBankOrNonAccountPage` Please suggest is it okay?

**Comment by parasharrajat** [comment]
The link is still incorrect. Try to follow the PR template. We have a big comment that explains it. 

Just replace your whole link line with the following:

`$ https://github.com/Expensify/App/issues/8541`

File: src/pages/workspace/WorkSpaceUtils.js (hunk 1 above)
Comment (parasharrajat) [reply src/pages/workspace/WorkSpaceUtils.js]: I don't see this suggestion applied.

File: src/pages/workspace/WorkSpaceUtils.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import lodashGet from 'lodash/get';
+import BankAccount from '../../libs/models/BankAccount';
+import Navigation from '../../libs/Navigation/Navigation';
+import ROUTES from '../../ROUTES';
+
+function getShouldShowBankOrNonAccountPage(reimbursementAccount, policyID) {
```
[the remark this answers is not in the document]
Comment (parasharrajat) [reply src/pages/workspace/WorkSpaceUtils.js:6]: No jsDOC Comment..s

File: src/pages/workspace/WorkspaceInitialPage.js [hunk 5]
```diff
@@ -121,7 +122,7 @@ class WorkspaceInitialPage extends React.Component {
             {
                 translationKey: 'workspace.common.bankAccount',
                 icon: Expensicons.Bank,
-                action: () => Navigation.navigate(ROUTES.getWorkspaceBankAccountRoute(policy.id)),
+                action: () => WorkSpaceUtils.getShouldShowBankOrNonAccountPage(this.props, ROUTES.getWorkspaceBankAccountRoute(policy.id), ROUTES.getBankAccountRoute()),
```
Comment (parasharrajat) [inline src/pages/workspace/WorkspaceInitialPage.js:125]: Is this call matching with the function signature?

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Request me, next review when you have tested your solution. There are a few syntax errors.

File: src/pages/workspace/WorkspaceInitialPage.js [hunk 6]
```diff
@@ -121,7 +122,7 @@ class WorkspaceInitialPage extends React.Component {
             {
                 translationKey: 'workspace.common.bankAccount',
                 icon: Expensicons.Bank,
-                action: () => Navigation.navigate(ROUTES.getWorkspaceBankAccountRoute(policy.id)),
+                action: () => navigateToBankAccountRoute(this.props, ROUTES.getWorkspaceBankAccountRoute(policy.id), ROUTES.getBankAccountRoute()),
```
[the remark this answers is not in the document]
Comment (dharmik) [reply src/pages/workspace/WorkspaceInitialPage.js]: Updated.

File: src/pages/workspace/WorkSpaceUtils.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import lodashGet from 'lodash/get';
+import BankAccount from '../../libs/models/BankAccount';
+import Navigation from '../../libs/Navigation/Navigation';
+import ROUTES from '../../ROUTES';
+
+function navigateToBankAccountRoute(reimbursementAccount, policyID) {
```
[the remark this answers is not in the document]
Comment (dharmik) [reply src/pages/workspace/WorkSpaceUtils.js:12]: Added.

File: src/pages/workspace/WorkspaceInitialPage.js (hunk 5 above)
Comment (dharmik) [reply src/pages/workspace/WorkspaceInitialPage.js]: Changes applied.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Please merge main. Sign all the commits.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Do we need https://github.com/Expensify/App/blob/503cd54a587dec6841adeabe69431d38d9293dd5/src/pages/workspace/WorkspaceBankAccountPage.js#L54 now?

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/pages/workspace/WorkSpaceUtils.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import Navigation from '../../libs/Navigation/Navigation';
+import ROUTES from '../../ROUTES';
+
+/**
+ * Navigate Bank Account Route
+ *
+ * @param {String} reimbursementAccount
+ * @param {String} policyID
+ */
+function navigateToBankAccountRoute(reimbursementAccount, policyID) {
```
[the remark this answers is not in the document]
Comment (dharmik) [reply src/pages/workspace/WorkSpaceUtils.js:12]: can we add function navigateToBankAccountRoute in `src/libs/Navigation/Navigation.js` and use in other file as

`action: () => Navigation.navigateToBankAccountRoute(this.props.reimbursementAccount, policy.id),`

Is it looks okay?

[the remark this answers is not in the document]
Comment (dharmik) [reply src/pages/workspace/WorkSpaceUtils.js:12]: > This method is an action and not util. Move this to ReimbursementAccount/navigation actions and then export it in `index.js` under the same. We can do better with the comment.

Sorry, Rajat may be you explain correct but I did not get it. Can you please help me to make it more.

**Comment by dharmik** [comment]
@parasharrajat Can you please review PR?
I tried to solve past unsigned commit but not fixed yet, will check it.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
QA steps are incomplete. 

> 4. Tap on connect bank account

What should happen after that? 

> Tap on issue card

What is issue card and how to navigate to it? Where are the QA steps for other pages you have touched?


This is still wrong :confused: 
image

Remove your line and paste this instead `$ https://github.com/Expensify/App/issues/8541`.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
PLease merge main..

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/pages/workspace/bills/WorkspaceBillsNoVBAView.js [hunk 9]
```diff
@@ -47,4 +49,11 @@ const WorkspaceBillsNoVBAView = props => (
...
 
-export default withLocalize(WorkspaceBillsNoVBAView);
+export default compose(
+    withLocalize,
+    withOnyx({
+        reimbursementAccount: {
+            key: ONYXKEYS.REIMBURSEMENT_ACCOUNT,
+        },
+    }),
+)(WorkspaceBillsNoVBAView);
```
[the remark this answers is not in the document]
Comment (dharmik) [reply src/pages/workspace/bills/WorkspaceBillsNoVBAView.js]: Resolved

File: src/pages/workspace/travel/WorkspaceTravelNoVBAView.js [hunk 10]
```diff
@@ -44,4 +46,11 @@ const WorkspaceTravelNoVBAView = props => (
...
 
-export default withLocalize(WorkspaceTravelNoVBAView);
+export default compose(
+    withLocalize,
+    withOnyx({
+        reimbursementAccount: {
+            key: ONYXKEYS.REIMBURSEMENT_ACCOUNT,
+        },
+    }),
+)(WorkspaceTravelNoVBAView);
```
[the remark this answers is not in the document]
Comment (dharmik) [reply src/pages/workspace/travel/WorkspaceTravelNoVBAView.js]: Resolved

File: src/pages/workspace/invoices/WorkspaceInvoicesNoVBAView.js [hunk 11]
```diff
@@ -47,4 +49,11 @@ const WorkspaceInvoicesNoVBAView = props => (
...
 
-export default withLocalize(WorkspaceInvoicesNoVBAView);
+export default compose(
+    withLocalize,
+    withOnyx({
+        reimbursementAccount: {
+            key: ONYXKEYS.REIMBURSEMENT_ACCOUNT,
+        },
+    }),
+)(WorkspaceInvoicesNoVBAView);
```
[the remark this answers is not in the document]
Comment (dharmik) [reply src/pages/workspace/invoices/WorkspaceInvoicesNoVBAView.js]: Resolved

**Comment by dharmik** [comment]
@parasharrajat Facing issue in signedCommits so close this PR and raised new one.

https://github.com/Expensify/App/pull/10456

**Comment by dharmik** [comment]
@parasharrajat PR Raised.

**Pull request closed without merging**